cpp/examples/arrow/dataset_skyhook_scan_example.cc

```diff
@@ -0,0 +1,192 @@
+
+  return dataset;
+}
+
+arrow::Result<std::shared_ptr<ds::Dataset>> GetDatasetFromFile(
+    std::shared_ptr<fs::FileSystem> fs, std::shared_ptr<ds::FileFormat> format,
+    std::string file) {
+  ds::FileSystemFactoryOptions options;
+  // The factory will try to build a child dataset.
+  ARROW_ASSIGN_OR_RAISE(auto factory,
+                        ds::FileSystemDatasetFactory::Make(fs, {file}, format, options));
+
+  // Try to infer a common schema for all files.
+  ARROW_ASSIGN_OR_RAISE(auto schema, factory->Inspect(kConf.inspect_options));
+  // Caller can optionally decide another schema as long as it is compatible
+  // with the previous one, e.g. `factory->Finish(compatible_schema)`.
+  ARROW_ASSIGN_OR_RAISE(auto child, factory->Finish(kConf.finish_options));
+
+  ds::DatasetVector children;
+  children.resize(kConf.repeat, child);
+  ARROW_ASSIGN_OR_RAISE(auto dataset,
+                        ds::UnionDataset::Make(std::move(schema), std::move(children)));
+
+  return dataset;
+}
+
+arrow::Result<std::shared_ptr<ds::Dataset>> GetDatasetFromPath(
+    std::shared_ptr<fs::FileSystem> fs, std::shared_ptr<ds::FileFormat> format,
+    std::string path) {
+  ARROW_ASSIGN_OR_RAISE(auto info, fs->GetFileInfo(path));
+  if (info.IsDirectory()) {
+    return GetDatasetFromDirectory(fs, format, path);
+  }
+  return GetDatasetFromFile(fs, format, path);
+}
+
+arrow::Result<std::shared_ptr<ds::Scanner>> GetScannerFromDataset(
+    std::shared_ptr<ds::Dataset> dataset, std::vector<std::string> columns,
+    cp::Expression filter, bool use_threads) {
+  ARROW_ASSIGN_OR_RAISE(auto scanner_builder, dataset->NewScan());
+
+  if (!columns.empty()) {
+    ABORT_ON_FAILURE(scanner_builder->Project(columns));
+  }
+
+  ABORT_ON_FAILURE(scanner_builder->Filter(filter));
+
+  ABORT_ON_FAILURE(scanner_builder->UseThreads(use_threads));
+
+  return scanner_builder->Finish();
+}
+
+arrow::Result<std::shared_ptr<Table>> GetTableFromScanner(
+    std::shared_ptr<ds::Scanner> scanner) {
+  return scanner->ToTable();
+}
+
+arrow::Result<std::shared_ptr<skyhook::SkyhookFileFormat>> InstantiateSkyhookFormat() {
+  std::string ceph_config_path = "/etc/ceph/ceph.conf";
+  std::string ceph_data_pool = "cephfs_data";
+  std::string ceph_user_name = "client.admin";
+  std::string ceph_cluster_name = "ceph";
+  std::string ceph_cls_name = "arrow";
+  std::shared_ptr<skyhook::RadosConnCtx> rados_ctx =
+      std::make_shared<skyhook::RadosConnCtx>(ceph_config_path, ceph_data_pool,
+                                              ceph_user_name, ceph_cluster_name,
+                                              ceph_cls_name);
+  ARROW_ASSIGN_OR_RAISE(auto format,
+                        skyhook::SkyhookFileFormat::Make(rados_ctx, "parquet"));
+  return format;
+}
+
+arrow::Status Main(char** argv) {
+  ARROW_ASSIGN_OR_RAISE(auto format, InstantiateSkyhookFormat());
+  std::string path;
+
+  ARROW_ASSIGN_OR_RAISE(auto fs, GetFileSystemFromUri(argv[1], &path));
```

Comment:
nit: use `Main(std::string dataset_root)` or something, and just call `Main(argv[1])` so that we aren't working with C-style strings here